pkg/install/vault/init.go

```diff
@@ -0,0 +1,212 @@
+	"fmt"
+
+	vaultapi "github.com/hashicorp/vault/api"
+	"github.com/hashicorp/vault/sdk/helper/consts"
+	"github.com/puppetlabs/leg/vaultutil/pkg/model"
+	vaultutil "github.com/puppetlabs/leg/vaultutil/pkg/vault"
+)
+
+type VaultInitializer struct {
+	vaultConfig        *vaultutil.VaultConfig
+	vaultCoreConfig    *VaultCoreConfig
+	vaultSystemManager model.VaultSystemManager
+}
+
+func (vi *VaultInitializer) InitializeVault(ctx context.Context) error {
+	credentials := &model.VaultKeys{
+		RootToken:  vi.vaultConfig.Token,
+		UnsealKeys: []string{vi.vaultConfig.UnsealKey},
+	}
+
+	var err error
+	if credentials.RootToken == "" || credentials.UnsealKeys[0] == "" {
+		credentials, err = vi.vaultSystemManager.GetCredentials(ctx)
+		if err != nil {
+			return err
+		}
+
+		if credentials == nil {
+			credentials, err = vi.vaultSystemManager.Initialize(ctx)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	err = vi.vaultSystemManager.Unseal(credentials)
+	if err != nil {
+		return err
+	}
+
+	vi.vaultSystemManager.SetToken(credentials)
+
+	err = vi.vaultSystemManager.EnableKubernetesAuth()
+	if err != nil {
+		return err
+	}
+
+	err = vi.vaultSystemManager.ConfigureKubernetesAuth(ctx)
+	if err != nil {
+		return err
+	}
+
+	err = vi.vaultSystemManager.EnableJWTAuth()
+	if err != nil {
+		return err
+	}
+
+	err = vi.vaultSystemManager.ConfigureJWTAuth(ctx)
+	if err != nil {
+		return err
+	}
+
+	// FIXME Upgrade to the latest oauthapp plugin
+	plugins := []*vaultapi.RegisterPluginInput{
+		{
+			Name:    "oauthapp",
+			Type:    consts.PluginTypeSecrets,
+			SHA256:  "2455ad9450e415efb42fd28c1436f4bf7e377524be5e534e55e6658b8ef56bd2",
+			Command: "vault-plugin-secrets-oauthapp-v3.0.0-beta.3-linux-amd64",
+		},
+	}
+
+	secretEngines := []*model.VaultSecretEngine{
+		{
+			Name: vi.vaultCoreConfig.LogServicePath,
+			Type: model.VaultSecretEngineTypeKVV2,
+		},
+		{
+			Name: vi.vaultCoreConfig.TenantPath,
+			Type: model.VaultSecretEngineTypeKVV2,
+		},
+		{
+			Name: vi.vaultCoreConfig.TransitPath,
+			Type: model.VaultSecretEngineTypeTransit,
+		},
+	}
+
+	policyGen := &vaultPolicyGenerator{
+		LogServicePath: vi.vaultCoreConfig.LogServicePath,
+		TenantPath:     vi.vaultCoreConfig.TenantPath,
+		TransitKey:     vi.vaultCoreConfig.TransitKey,
+		TransitPath:    vi.vaultCoreConfig.TransitPath,
+	}
+
+	policies := []*model.VaultPolicy{
+		{
+			Name:  vi.vaultCoreConfig.LogServiceVaultAgentRole,
+			Rules: string(policyGen.logServicePolicy()),
+		},
+		{
+			Name:  vi.vaultCoreConfig.MetadataAPIVaultAgentRole,
+			Rules: string(policyGen.metadataAPIPolicy()),
+		},
+		{
+			Name:  vi.vaultCoreConfig.OperatorVaultAgentRole,
+			Rules: string(policyGen.operatorPolicy()),
+		},
+	}
+
+	auth, err := vi.vaultSystemManager.GetAuthMethod(fmt.Sprintf("%s/", vi.vaultConfig.JWTMount))
+	if err != nil {
+		return err
+	}
+
+	if auth != nil {
+		policyGen.AuthJWTAccessor = auth.Accessor
+
+		// FIXME Add installer configuration for `metadata-api-tenant`
```

Comment:
Same for this -- seems like this is done? Anything else that needs to be taken care of here?

Reply:
Same as above. The logic/functionality works fine. But ideally we make this naming configurable, similar to other roles/policies/etc.

Reply:
To that point, though, it's not really a FIXME any longer and should be transitioned to TODO.